// @spartacus/schematics/src/add-spartacus/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const terminal_1 = require("@angular-devkit/core/src/terminal");
const schematics_1 = require("@angular-devkit/schematics");
const tasks_1 = require("@angular-devkit/schematics/tasks");
const schematics_2 = require("@angular/cdk/schematics");
const ast_utils_1 = require("@schematics/angular/utility/ast-utils");
const dependencies_1 = require("@schematics/angular/utility/dependencies");
const ng_ast_utils_1 = require("@schematics/angular/utility/ng-ast-utils");
const project_targets_1 = require("@schematics/angular/utility/project-targets");
const constants_1 = require("../shared/constants");
const file_utils_1 = require("../shared/utils/file-utils");
const module_file_utils_1 = require("../shared/utils/module-file-utils");
const package_utils_1 = require("../shared/utils/package-utils");
const transform_utils_1 = require("../shared/utils/transform-utils");
const workspace_utils_1 = require("../shared/utils/workspace-utils");
function addPackageJsonDependencies() {
    return (tree, context) => {
        const spartacusVersion = `^${package_utils_1.getSpartacusSchematicsVersion()}`;
        const ngrxVersion = '~9.1.0';
        const angularVersion = package_utils_1.getAngularVersion(tree);
        const dependencies = [
            {
                type: dependencies_1.NodeDependencyType.Default,
                version: spartacusVersion,
                name: constants_1.SPARTACUS_CORE,
            },
            {
                type: dependencies_1.NodeDependencyType.Default,
                version: spartacusVersion,
                name: constants_1.SPARTACUS_STOREFRONTLIB,
            },
            {
                type: dependencies_1.NodeDependencyType.Default,
                version: spartacusVersion,
                name: constants_1.SPARTACUS_ASSETS,
            },
            {
                type: dependencies_1.NodeDependencyType.Default,
                version: spartacusVersion,
                name: constants_1.SPARTACUS_STYLES,
            },
            {
                type: dependencies_1.NodeDependencyType.Default,
                version: '^6.0.0',
                name: '@ng-bootstrap/ng-bootstrap',
            },
            {
                type: dependencies_1.NodeDependencyType.Default,
                version: '^4.0.0',
                name: '@ng-select/ng-select',
            },
            {
                type: dependencies_1.NodeDependencyType.Default,
                version: ngrxVersion,
                name: '@ngrx/store',
            },
            {
                type: dependencies_1.NodeDependencyType.Default,
                version: ngrxVersion,
                name: '@ngrx/effects',
            },
            {
                type: dependencies_1.NodeDependencyType.Default,
                version: ngrxVersion,
                name: '@ngrx/router-store',
            },
            {
                type: dependencies_1.NodeDependencyType.Default,
                version: '^4.2.1',
                name: 'bootstrap',
            },
            { type: dependencies_1.NodeDependencyType.Default, version: '^19.3.4', name: 'i18next' },
            {
                type: dependencies_1.NodeDependencyType.Default,
                version: '^3.2.2',
                name: 'i18next-xhr-backend',
            },
            {
                type: dependencies_1.NodeDependencyType.Default,
                version: angularVersion,
                name: '@angular/service-worker',
            },
            {
                type: dependencies_1.NodeDependencyType.Default,
                version: angularVersion,
                name: constants_1.ANGULAR_LOCALIZE,
            },
            {
                type: dependencies_1.NodeDependencyType.Default,
                version: '^8.0.0',
                name: 'ngx-infinite-scroll',
            },
        ];
        dependencies.forEach((dependency) => {
            dependencies_1.addPackageJsonDependency(tree, dependency);
            context.logger.info(`âœ…ï¸ Added '${dependency.name}' into ${dependency.type}`);
        });
        return tree;
    };
}
function installPackageJsonDependencies() {
    return (tree, context) => {
        context.addTask(new tasks_1.NodePackageInstallTask());
        context.logger.log('info', `ðŸ” Installing packages...`);
        return tree;
    };
}
function prepareSiteContextConfig(options) {
    const currency = transform_utils_1.parseCSV(options.currency, ['USD']).toUpperCase();
    const language = transform_utils_1.parseCSV(options.language, ['en']).toLowerCase();
    let context = `
      context: {
        currency: [${currency}],
        language: [${language}],`;
    if (options.baseSite) {
        const baseSites = transform_utils_1.parseCSV(options.baseSite);
        context += `
        baseSite: [${baseSites}]`;
    }
    context += `
      },`;
    return context;
}
function getStorefrontConfig(options) {
    const baseUrlPart = `\n          baseUrl: '${options.baseUrl}',`;
    const context = prepareSiteContextConfig(options);
    return `{
      backend: {
        occ: {${options.useMetaTags ? '' : baseUrlPart}
          prefix: '${options.occPrefix}'
        }
      },${context}
      i18n: {
        resources: translations,
        chunks: translationChunksConfig,
        fallbackLang: 'en'
      },
      features: {
        level: '${options.featureLevel || package_utils_1.getSpartacusCurrentFeatureLevel()}'
      }
    }`;
}
function updateAppModule(options) {
    return (host, context) => {
        context.logger.debug('Updating main module');
        // find app module
        const projectTargets = project_targets_1.getProjectTargets(host, options.project);
        if (!projectTargets.build) {
            throw new schematics_1.SchematicsException(`Project target "build" not found.`);
        }
        const mainPath = projectTargets.build.options.main;
        const modulePath = ng_ast_utils_1.getAppModulePath(host, mainPath);
        context.logger.debug(`main module path: ${modulePath}`);
        const moduleSource = file_utils_1.getTsSourceFile(host, modulePath);
        if (!ast_utils_1.isImported(moduleSource, constants_1.B2C_STOREFRONT_MODULE, constants_1.SPARTACUS_STOREFRONTLIB)) {
            // add imports
            module_file_utils_1.addImport(host, modulePath, 'translations', constants_1.SPARTACUS_ASSETS);
            module_file_utils_1.addImport(host, modulePath, 'translationChunksConfig', constants_1.SPARTACUS_ASSETS);
            module_file_utils_1.addImport(host, modulePath, constants_1.B2C_STOREFRONT_MODULE, constants_1.SPARTACUS_STOREFRONTLIB);
            module_file_utils_1.addToModuleImportsAndCommitChanges(host, modulePath, `${constants_1.B2C_STOREFRONT_MODULE}.withConfig(${getStorefrontConfig(options)})`);
        }
        return host;
    };
}
function installStyles(project) {
    return (host) => {
        const styleFilePath = schematics_2.getProjectStyleFile(project);
        if (!styleFilePath) {
            console.warn(terminal_1.red(`Could not find the default style file for this project.`));
            console.warn(terminal_1.red(`Please consider manually setting up spartacus styles`));
            return;
        }
        if (styleFilePath.split('.').pop() !== 'scss') {
            console.warn(terminal_1.red(`Could not find the default SCSS style file for this project. `));
            console.warn(terminal_1.red(`Please make sure your project is configured with SCSS and consider manually setting up spartacus styles.`));
            return;
        }
        const buffer = host.read(styleFilePath);
        if (!buffer) {
            console.warn(terminal_1.red(`Could not read the default style file within the project ` +
                `(${terminal_1.italic(styleFilePath)})`));
            console.warn(terminal_1.red(`Please consider manually importing spartacus styles.`));
            return;
        }
        const htmlContent = buffer.toString();
        const insertion = '\n' + `@import '~@spartacus/styles/index';\n`;
        if (htmlContent.includes(insertion)) {
            return;
        }
        const recorder = host.beginUpdate(styleFilePath);
        recorder.insertLeft(htmlContent.length, insertion);
        host.commitUpdate(recorder);
    };
}
function updateMainComponent(project, options) {
    return (host, _context) => {
        const filePath = project.sourceRoot + '/app/app.component.html';
        const buffer = host.read(filePath);
        if (!buffer) {
            console.warn(terminal_1.red(`Could not read app.component.html file.`));
            return;
        }
        const htmlContent = buffer.toString();
        const insertion = `<cx-storefront></cx-storefront>\n`;
        if (htmlContent.includes(insertion)) {
            return;
        }
        const recorder = host.beginUpdate(filePath);
        if (options && options.overwriteAppComponent) {
            recorder.remove(0, htmlContent.length);
            recorder.insertLeft(0, insertion);
        }
        else {
            recorder.insertLeft(htmlContent.length, `\n${insertion}`);
        }
        host.commitUpdate(recorder);
        return host;
    };
}
function updateIndexFile(project, options) {
    return (host) => {
        const projectIndexHtmlPath = file_utils_1.getIndexHtmlPath(project);
        const baseUrl = options.baseUrl || 'OCC_BACKEND_BASE_URL_VALUE';
        const metaTags = [
            `<meta name="occ-backend-base-url" content="${baseUrl}" />`,
            `<meta name="media-backend-base-url" content="MEDIA_BACKEND_BASE_URL_VALUE" />`,
        ];
        metaTags.forEach((metaTag) => {
            schematics_2.appendHtmlElementToHead(host, projectIndexHtmlPath, metaTag);
        });
        return host;
    };
}
function addSpartacus(options) {
    return (tree, context) => {
        const project = workspace_utils_1.getProjectFromWorkspace(tree, options);
        return schematics_1.chain([
            addPackageJsonDependencies(),
            updateAppModule(options),
            installStyles(project),
            updateMainComponent(project, options),
            options.useMetaTags ? updateIndexFile(project, options) : schematics_1.noop(),
            installPackageJsonDependencies(),
        ])(tree, context);
    };
}
exports.addSpartacus = addSpartacus;
