// @spartacus/schematics/src/shared/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UTF_8 = 'utf-8';
/***** Libraries versions start *****/
exports.DEFAULT_ANGULAR_VERSION = '^9.1.0';
exports.ANGULAR_UNIVERSAL_EXPRESS_VERSION = '^9.1.0';
exports.ANGULAR_UNIVERSAL_BUILDERS = '^9.1.0';
/***** Libraries versions end *****/
/***** Imports start *****/
exports.ANGULAR_SCHEMATICS = '@schematics/angular';
exports.ANGULAR_CORE = '@angular/core';
exports.ANGULAR_FORMS = '@angular/forms';
exports.ANGULAR_ROUTER = '@angular/router';
exports.ANGULAR_HTTP = '@angular/common/http';
exports.ANGULAR_LOCALIZE = '@angular/localize';
exports.RXJS = 'rxjs';
exports.ANGULAR_COMMON = '@angular/common';
exports.ANGULAR_PLATFORM_BROWSER = '@angular/platform-browser';
exports.SPARTACUS_CORE = '@spartacus/core';
exports.SPARTACUS_STOREFRONTLIB = '@spartacus/storefront';
exports.SPARTACUS_CDS = '@spartacus/cds';
exports.SPARTACUS_ASSETS = '@spartacus/assets';
exports.SPARTACUS_STYLES = '@spartacus/styles';
exports.NGRX_STORE = '@ngrx/store';
/***** Imports end *****/
/***** Classes start *****/
exports.OBSERVABLE_CLASS = 'Observable';
exports.RENDERER_2 = 'Renderer2';
exports.ELEMENT_REF = 'ElementRef';
exports.INJECTOR = 'Injector';
exports.STORE = 'Store';
exports.FORM_BUILDER = 'FormBuilder';
exports.CHANGE_DETECTOR_REF = 'ChangeDetectorRef';
exports.ACTIVATED_ROUTE = 'ActivatedRoute';
exports.ROUTER = 'Router';
exports.WINDOW_REF = 'WindowRef';
exports.VIEW_CONTAINER_REF = 'ViewContainerRef';
exports.TEMPLATE_REF = 'TemplateRef';
exports.LOCATION = 'Location';
exports.COMPONENT_FACTORY_RESOLVER = 'ComponentFactoryResolver';
exports.HTTP_CLIENT = 'HttpClient';
exports.B2C_STOREFRONT_MODULE = 'B2cStorefrontModule';
exports.CMS_COMPONENT_DATA_CLASS = 'CmsComponentData';
exports.CONFIG_MODULE_CLASS = 'ConfigModule';
exports.CMS_CONFIG = 'CmsConfig';
exports.CMS_SELECTORS = 'CmsSelectors';
exports.CMS_ACTIONS = 'CmsActions';
exports.LOAD_CMS_COMPONENT_CLASS = 'LoadCmsComponent';
exports.LOAD_CMS_COMPONENT_FAIL_CLASS = 'LoadCmsComponentFail';
exports.LOAD_CMS_COMPONENT_SUCCESS_CLASS = 'LoadCmsComponentSuccess';
exports.CMS_GET_COMPONENT_FROM_PAGE = 'CmsGetComponentFromPage';
exports.USER_ADDRESS_SERVICE = 'UserAddressService';
exports.AUTH_SERVICE = 'AuthService';
exports.FEATURE_CONFIG_SERVICE = 'FeatureConfigService';
exports.SEMANTIC_PATH_SERVICE = 'SemanticPathService';
exports.CMS_SERVICE = 'CmsService';
exports.PAGE_META_SERVICE = 'PageMetaService';
exports.STORE_DATA_SERVICE = 'StoreDataService';
exports.PAGE_META_RESOLVER = 'PageMetaResolver';
exports.PRODUCT_PAGE_META_RESOLVER = 'ProductPageMetaResolver';
exports.CATEGORY_PAGE_META_RESOLVER = 'CategoryPageMetaResolver';
exports.CHECKOUT_PAGE_META_RESOLVER = 'CheckoutPageMetaResolver';
exports.PRODUCT_SEARCH_SERVICE = 'ProductSearchService';
exports.CHECKOUT_SERVICE = 'CheckoutService';
exports.CHECKOUT_PAYMENT_SERVICE = 'CheckoutPaymentService';
exports.CHECKOUT_DELIVERY_SERVICE = 'CheckoutDeliveryService';
exports.CART_DATA_SERVICE = 'CartDataService';
exports.CART_SERVICE = 'CartService';
exports.ACTIVE_CART_SERVICE = 'ActiveCartService';
exports.PROMOTION_SERVICE = 'PromotionService';
exports.ORDER_DETAILS_SERVICE = 'OrderDetailsService';
exports.CHECKOUT_LOGIN_COMPONENT = 'CheckoutLoginComponent';
exports.AUTH_REDIRECT_SERVICE = 'AuthRedirectService';
exports.CHECKOUT_DETAILS_SERVICE = 'CheckoutDetailsService';
exports.NOT_CHECKOUT_AUTH_GUARD = 'NotCheckoutAuthGuard';
exports.ROUTING_SERVICE = 'RoutingService';
exports.SHIPPING_ADDRESS_COMPONENT = 'ShippingAddressComponent';
exports.CHECKOUT_CONFIG_SERVICE = 'CheckoutConfigService';
exports.TRANSLATION_SERVICE = 'TranslationService';
exports.ADD_TO_CART_COMPONENT = 'AddToCartComponent';
exports.ADDED_TO_CART_DIALOG_COMPONENT = 'AddedToCartDialogComponent';
exports.MODAL_SERVICE = 'ModalService';
exports.CURRENT_PRODUCT_SERVICE = 'CurrentProductService';
exports.CART_NOT_EMPTY_GUARD = 'CartNotEmptyGuard';
exports.CART_TOTALS_COMPONENT = 'CartTotalsComponent';
exports.MINI_CART_COMPONENT = 'MiniCartComponent';
exports.CHECKOUT_ORDER_SUMMARY_COMPONENT = 'CheckoutOrderSummaryComponent';
exports.CHECKOUT_PROGRESS_MOBILE_TOP_COMPONENT = 'CheckoutProgressMobileTopComponent';
exports.CHECKOUT_CONFIG = 'CheckoutConfig';
exports.ROUTING_CONFIG_SERVICE = 'RoutingConfigService';
exports.PAYMENT_METHOD_COMPONENT = 'PaymentMethodComponent';
exports.USER_PAYMENT_SERVICE = 'UserPaymentService';
exports.GLOBAL_MESSAGE_SERVICE = 'GlobalMessageService';
exports.CHECKOUT_AUTH_GUARD = 'CheckoutAuthGuard';
exports.CART_PAGE_LAYOUT_HANDLER = 'CartPageLayoutHandler';
exports.SELECTIVE_CART_SERVICE = 'SelectiveCartService';
exports.CDS_SPARTACUS_EVENT_SERVICE = 'SpartacusEventService';
exports.CONSENT_SERVICE = 'ConsentService';
exports.CDS_CONFIG = 'CdsConfig';
exports.ADDRESS_BOOK_COMPONENT_SERVICE = 'AddressBookComponentService';
exports.CHECKOUT_GUARD = 'CheckoutGuard';
exports.LOGOUT_GUARD = 'LogoutGuard';
exports.EXPRESS_CHECKOUT_SERVICE = 'ExpressCheckoutService';
exports.CONSENT_MANAGEMENT_FORM_COMPONENT = 'ConsentManagementFormComponent';
exports.CLIENT_AUTHENTICATION_TOKEN_SERVICE = 'ClientAuthenticationTokenService';
exports.USER_AUTHENTICATION_TOKEN_SERVICE = 'UserAuthenticationTokenService';
exports.OCC_ENDPOINTS_SERVICE = 'OccEndpointsService';
exports.AUTH_CONFIG = 'AuthConfig';
exports.OCC_CART_ENTRY_ADAPTER = 'OccCartEntryAdapter';
exports.OCC_CART_ADAPTER = 'OccCartAdapter';
exports.OCC_USER_ORDER_ADAPTER = 'OccUserOrderAdapter';
exports.CONVERTER_SERVICE = 'ConverterService';
exports.CONSENT_MANAGEMENT_COMPONENT = 'ConsentManagementComponent';
exports.ANONYMOUS_CONSENT_DIALOG_COMPONENT = 'AnonymousConsentDialogComponent';
exports.PRODUCT_IMAGES_COMPONENT = 'ProductImagesComponent';
exports.USER_CONSENT_SERVICE = 'UserConsentService';
exports.USER_ORDER_SERVICE = 'UserOrderService';
exports.USER_SERVICE = 'UserService';
exports.CART_DETAILS_COMPONENT = 'CartDetailsComponent';
exports.ORDER_DETAIL_ITEMS_COMPONENT = 'OrderDetailItemsComponent';
exports.ORDER_CONFIRMATION_ITEMS_COMPONENT = 'OrderConfirmationItemsComponent';
exports.REVIEW_SUBMIT_COMPONENT = 'ReviewSubmitComponent';
exports.CART_VOUCHER_SERVICE = 'CartVoucherService';
exports.CART_COUPON_COMPONENT = 'CartCouponComponent';
exports.CUSTOMER_COUPON_SERVICE = 'CustomerCouponService';
exports.CART_ITEM_LIST_COMPONENT = 'CartItemListComponent';
exports.CART_ITEM_COMPONENT = 'CartItemComponent';
exports.PROTECTED_ROUTES_SERVICE = 'ProtectedRoutesService';
exports.LOGIN_FORM_COMPONENT = 'LoginFormComponent';
exports.REGISTER_COMPONENT = 'RegisterComponent';
exports.ANONYMOUS_CONSENTS_SERVICE = 'AnonymousConsentsService';
exports.ANONYMOUS_CONSENTS_CONFIG = 'AnonymousConsentsConfig';
exports.STAR_RATING_COMPONENT = 'StarRatingComponent';
exports.PRODUCT_SERVICE = 'ProductService';
exports.PRODUCT_LOADING_SERVICE = 'ProductLoadingService';
exports.PRODUCT_CAROUSEL_COMPONENT = 'ProductCarouselComponent';
exports.PRODUCT_SCROLL_COMPONENT = 'ProductScrollComponent';
exports.PRODUCT_LIST_COMPONENT = 'ProductListComponent';
exports.PRODUCT_LIST_COMPONENT_SERVICE = 'ProductListComponentService';
exports.PAGE_LAYOUT_SERVICE = 'PageLayoutService';
exports.VIEW_CONFIG = 'ViewConfig';
exports.FOOTER_NAVIGATION_COMPONENT = 'FooterNavigationComponent';
exports.NAVIGATION_SERVICE = 'NavigationService';
exports.URL_MATCHER_SERVICE = 'UrlMatcherService';
exports.CONFIGURABLE_ROUTES_SERVICE = 'ConfigurableRoutesService';
exports.EXTERNAL_ROUTES_SERVICE = 'ExternalRoutesService';
exports.EXTERNAL_ROUTES_CONFIG = 'ExternalRoutesConfig';
exports.PAGE_SLOT_COMPONENT = 'PageSlotComponent';
exports.DYNAMIC_ATTRIBUTE_SERVICE = 'DynamicAttributeService';
exports.OUTLET_DIRECTIVE = 'OutletDirective';
exports.OUTLET_SERVICE = 'OutletService';
exports.DEFER_LOADER_SERVICE = 'DeferLoaderService';
exports.LAUNCH_DIALOG_SERVICE = 'LaunchComponentService';
exports.PLACE_ORDER_COMPONENT = 'PlaceOrderComponent';
exports.USER_INTERESTS_SERVICE = 'UserInterestsService';
exports.USER_NOTIFICATION_PREFERENCE_SERVICE = 'UserNotificationPreferenceService';
exports.PRODUCT_REVIEWS_COMPONENT = `ProductReviewsComponent`;
exports.SEARCH_BOX_COMPONENT = `SearchBoxComponent`;
exports.ADDRESS_BOOK_COMPONENT = `AddressBookComponent`;
exports.PAYMENT_FORM_COMPONENT = 'PaymentFormComponent';
exports.AMEND_ORDER_ACTIONS_COMPONENT = 'AmendOrderActionsComponent';
exports.COMPONENT_WRAPPER_DIRECTIVE = 'ComponentWrapperDirective';
exports.CMS_MAPPING_SERVICE = 'CmsMappingService';
exports.COMPONENT_HANDLER_SERVICE = 'ComponentHandlerService';
exports.CMS_INJECTOR_SERVICE = 'CmsInjectorService';
exports.SMART_EDIT_SERVICE = 'SmartEditService';
exports.STORE_FINDER_SEARCH_RESULT_COMPONENT = 'StoreFinderSearchResultComponent';
exports.STORE_FINDER_SERVICE = 'StoreFinderService';
exports.STORE_FINDER_CONFIG = 'StoreFinderConfig';
exports.PRODUCT_REVIEW_SERVICE = 'ProductReviewService';
exports.SEARCH_BOX_COMPONENT_SERVICE = 'SearchBoxComponentService';
exports.TAB_PARAGRAPH_CONTAINER_COMPONENT = 'TabParagraphContainerComponent';
exports.NAVIGATION_UI_COMPONENT = 'NavigationUIComponent';
exports.STORE_FINDER_LIST_ITEM_COMPONENT = 'StoreFinderListItemComponent';
exports.CART_CONFIG_SERVICE = 'CartConfigService';
exports.MULTI_CART_SERVICE = 'MultiCartService';
exports.BASE_SITE_SERVICE = 'BaseSiteService';
exports.QUALTRICS_LOADER_SERVICE = 'QualtricsLoaderService';
exports.RENDERER_FACTORY_2 = 'RendererFactory2';
exports.QUALTRICS_CONFIG = 'QualtricsConfig';
exports.QUALTRICS_COMPONENT = 'QualtricsComponent';
exports.PRODUCT_FACET_NAVIGATION_COMPONENT = 'ProductFacetNavigationComponent';
exports.BREAKPOINT_SERVICE = 'BreakpointService';
exports.STORE_FINDER_GRID_COMPONENT = 'StoreFinderGridComponent';
exports.OUTLET_RENDERER_SERVICE = 'OutletRendererService';
exports.SKIP_LINK_SERVICE = 'SkipLinkService';
exports.SKIP_LINK_CONFIG = 'SkipLinkConfig';
exports.KEYBOARD_FOCUS_SERVICE = 'KeyboardFocusService';
exports.STOREFRONT_COMPONENT = 'StorefrontComponent';
exports.HAMBURGER_MENU_SERVICE = 'HamburgerMenuService';
exports.ANONYMOUS_CONSENT_MANAGEMENT_BANNER_COMPONENT = 'AnonymousConsentManagementBannerComponent';
exports.ANONYMOUS_CONSENT_LAUNCH_DIALOG_SERVICE = 'AnonymousConsentLaunchDialogService';
exports.ANONYMOUS_CONSENT_OPEN_DIALOG_COMPONENT = 'AnonymousConsentOpenDialogComponent';
/***** Classes end *****/
/***** Removed public api start *****/
exports.CART_EFFECTS = 'CartEffects';
exports.WISHLIST_EFFECTS = 'WishlistEffects';
exports.CART_VOUCHER_EFFECTS = 'CartVoucherEffects';
exports.CART_ENTRY_EFFECTS = 'CartEntryEffects';
exports.CART_COMBINED_EFFECTS = 'effects';
exports.CART_SELECTORS = 'CartSelectors';
exports.CART_DATA_CONST = 'CART_DATA';
exports.STATE_WITH_CART = 'StateWithCart';
exports.CARTS_STATE = 'CartsState';
exports.CART_FEATURE_CONST = 'CART_FEATURE';
exports.CART_STATE = 'CartState';
exports.GET_REDUCERS = 'getReducers';
exports.REDUCER_TOKEN = 'reducerToken';
exports.REDUCER_PROVIDER = 'reducerProvider';
exports.CLEAR_CART_STATE = 'clearCartState';
exports.META_REDUCERS = 'metaReducers';
exports.CLEAR_MULTI_CART_STATE = 'clearMultiCartState';
exports.MULTI_CART_META_REDUCERS = 'multiCartMetaReducers';
exports.MULTI_CART_REDUCER_TOKEN = 'multiCartReducerToken';
exports.GET_MULTI_CART_REDUCERS = 'getMultiCartReducers';
exports.MULTI_CART_REDUCER_PROVIDER = 'multiCartReducerProvider';
exports.SITE_CONTEXT_PARAMS_PROVIDERS = 'siteContextParamsProviders';
exports.INIT_SITE_CONTEXT_ROUTES_HANDLER = 'initSiteContextRoutesHandler';
exports.INITITIALIZE_CONTEXT = 'inititializeContext';
exports.CONTEXT_SERVICE_PROVIDERS = 'contextServiceProviders';
exports.URL_MATCHER_FACTORY_SERVICE = 'UrlMatcherFactoryService';
exports.CMS_COMPONENTS_SERVICE = 'CmsComponentsService';
exports.ANONYMOUS_USERID_CONST = 'ANONYMOUS_USERID';
exports.PWA_CONFIGURATION_FACTORY = 'pwaConfigurationFactory';
exports.PWA_FACTORY = 'pwaFactory';
exports.GET_STRUCTURED_DATA_FACTORY = 'getStructuredDataFactory';
exports.SKIP_LINK_FACTORY = 'skipLinkFactory';
exports.FORM_UTILS = 'FormUtils';
exports.BILLING_ADDRESS_FORM_COMPONENT = 'BillingAddressFormComponent';
exports.BILLING_ADDRESS_FORM_MODULE = 'BillingAddressFormModule';
exports.LOADER_STATE = 'LoaderState';
exports.PROCESSES_LOADER_STATE = 'ProcessesLoaderState';
exports.ENTITY_STATE = 'EntityState';
exports.ENTITY_LOADER_STATE = 'EntityLoaderState';
exports.ENTITY_PROCESSES_LOADER_STATE = 'EntityProcessesLoaderState';
exports.STATE_LOADER_ACTIONS = 'StateLoaderActions';
exports.STATE_PROCESSES_LOADER_ACTIONS = 'StateProcessesLoaderActions';
exports.STATE_ENTITY_ACTIONS = 'StateEntityActions';
exports.STATE_ENTITY_LOADER_ACTIONS = 'StateEntityLoaderActions';
exports.STATE_ENTITY_PROCESSESS_LOADER_ACTIONS = 'StateEntityProcessesLoaderActions';
exports.STATE_LOADER_SELECTORS = 'StateLoaderSelectors';
exports.STATE_PROCESSES_LOADER_SELECTORS = 'StateProcessesLoaderSelectors';
exports.STATE_ENTITY_SELECTORS = 'StateEntitySelectors';
exports.STATE_ENTITY_LOADER_SELECTORS = 'StateEntityLoaderSelectors';
exports.STATE_ENTITY_PROCESSESS_LOADER_SELECTORS = 'StateEntityProcessesLoaderSelectors';
exports.LOADER_REDUCER = 'loaderReducer';
exports.PROCESSES_LOADER_REDUCER = 'processesLoaderReducer';
exports.ENTITY_REDUCER = 'entityReducer';
exports.ENTITY_LOADER_REDUCER = 'entityLoaderReducer';
exports.ENTITY_PROCESSES_LOADER_REDUCER = 'entityProcessesLoaderReducer';
exports.INITIAL_ENTITY_STATE = 'initialEntityState';
exports.INITIAL_LOADER_STATE = 'initialLoaderState';
exports.INITIAL_PROCESSES_STATE = 'initialProcessesState';
exports.GET_STATE_SLICE = 'getStateSlice';
exports.OF_LOADER_LOAD = 'ofLoaderLoad';
exports.OF_LOADER_FAIL = 'ofLoaderFail';
exports.OF_LOADER_SUCCESS = 'ofLoaderSuccess';
exports.ONLY_NUMBER_DIRECTIVE_MODULE = 'OnlyNumberDirectiveModule';
exports.ONLY_NUMBER_DIRECTIVE = 'OnlyNumberDirective';
exports.AUTO_FOCUS_DIRECTIVE = 'AutoFocusDirective';
exports.AUTO_FOCUS_DIRECTIVE_MODULE = 'AutoFocusDirectiveModule';
/***** Removed public api end *****/
/***** Properties start *****/
exports.CMS_COMPONENT_DATA_PROPERTY_NAME = 'componentData';
exports.KYMA_ENABLED = 'kyma_enabled';
exports.ANONYMOUS_CONSENTS = 'anonymousConsents';
/***** Properties end *****/
/***** APIs start *****/
exports.GET_COMPONENT_STATE_OLD_API = 'getComponentState';
exports.GET_COMPONENTS_STATE_NEW_API = 'getComponentsState';
exports.GET_COMPONENT_ENTITIES_OLD_API = 'getComponentEntities';
exports.COMPONENT_STATE_SELECTOR_FACTORY_OLD_API = 'componentStateSelectorFactory';
exports.COMPONENTS_STATE_SELECTOR_FACTORY_NEW_API = 'componentsLoaderStateSelectorFactory';
exports.COMPONENT_SELECTOR_FACTORY_OLD_API = 'componentSelectorFactory';
exports.COMPONENTS_SELECTOR_FACTORY_NEW_API = 'componentsSelectorFactory';
exports.IS_LAUNCH_IN_SMART_EDIT = 'isLaunchInSmartEdit';
exports.IS_LAUNCHED_IN_SMART_EDIT = 'isLaunchedInSmartEdit';
exports.ADD_DYNAMIC_ATTRIBUTES = 'addDynamicAttributes';
exports.GET_FALSY_URL_MATCHER = 'getFalsyUrlMatcher';
exports.GET_MULTIPLE_PATHS_URL_MATCHER = 'getMultiplePathsUrlMatcher';
exports.GET_PATH_URL_MATCHER = 'getPathUrlMatcher';
exports.GET_OPPOSITE_URL_MATCHER = 'getOppositeUrlMatcher';
exports.GET_GLOB_URL_MATCHER = 'getGlobUrlMatcher';
/***** APIs end *****/
/***** Misc start *****/
exports.TODO_SPARTACUS = 'TODO:Spartacus -';
exports.CSS_DOCS_URL = 'https://sap.github.io/spartacus-docs/updating-to-version-2/css';
/***** Misc end *****/
